packages/contracts-bedrock/scripts/deploy/UpgradeAnchorStateRegistry.s.sol

```diff
@@ -0,0 +1,191 @@
+pragma solidity ^0.8.15;
+
+// Forge
+import {Script} from "forge-std/Script.sol";
+import {console} from "forge-std/console.sol";
+
+// Scripts
+import {DeployUtils} from "scripts/libraries/DeployUtils.sol";
+
+// Libraries
+import {GameType, Hash, OutputRoot} from "src/dispute/lib/Types.sol";
+// Contracts
+import {StorageSetter} from "src/universal/StorageSetter.sol";
+
+// Interfaces
+import {IAnchorStateRegistry} from "interfaces/dispute/IAnchorStateRegistry.sol";
+import {IDisputeGameFactory} from "interfaces/dispute/IDisputeGameFactory.sol";
+import {IAnchorStateRegistry} from "interfaces/dispute/IAnchorStateRegistry.sol";
+import {IProxyAdmin} from "interfaces/universal/IProxyAdmin.sol";
+import {ISuperchainConfig} from "interfaces/L1/ISuperchainConfig.sol";
+
+/// @title UpgradeAnchorStateRegistry
+contract UpgradeAnchorStateRegistry is Script {
+    function run(
+        address _disputeGameFactoryProxy,
+        address _opChainProxyAdmin,
+        address _anchorStateRegistryProxy,
+        address _superchainConfig,
+        uint32 _type,
+        uint256 _l2BlockNumber,
+        bytes32 _outputRoot
+    ) public {
+        console.log("_disputeGameFactoryProxy: %s", _disputeGameFactoryProxy);
+        console.log("_opChainProxyAdmin: %s", _opChainProxyAdmin);
+        console.log("_anchorStateRegistryProxy: %s", _anchorStateRegistryProxy);
+        console.log("_superchainConfig: %s", _superchainConfig);
+        console.log("_type: %s", _type);
+        console.log("_l2BlockNumber: %s", _l2BlockNumber);
+        console.log("_outputRoot: %s", bytes32ToHex(_outputRoot));
+
+        vm.startBroadcast();
+        upgradeAnchorStateRegistryImpl(
+            IDisputeGameFactory(_disputeGameFactoryProxy),
+            IProxyAdmin(_opChainProxyAdmin),
+            IAnchorStateRegistry(_anchorStateRegistryProxy),
+            ISuperchainConfig(_superchainConfig),
+            GameType.wrap(_type),
+            _l2BlockNumber,
+            Hash.wrap(_outputRoot)
+        );
+        vm.stopBroadcast();
+        checkOutput(
+            IAnchorStateRegistry(_anchorStateRegistryProxy),
+            GameType.wrap(_type),
+            _l2BlockNumber,
+            Hash.wrap(_outputRoot)
+        );
+    }
+
+    function upgradeAnchorStateRegistryImpl(
+        IDisputeGameFactory _disputeGameFactoryProxy,
+        IProxyAdmin _opChainProxyAdmin,
+        IAnchorStateRegistry _anchorStateRegistryProxy,
+        ISuperchainConfig _superchainConfig,
+        GameType _type,
+        uint256 _l2BlockNumber,
+        Hash _outputRoot
+    ) internal {
+        address anchorStateRegistryImpl = DeployUtils.create1({
+            _name: "AnchorStateRegistry",
+            _args: abi.encode(_disputeGameFactoryProxy)
+        });
+
+        address storageSetter = DeployUtils.create1({
+            _name: "StorageSetter",
+            _args: ""
+        });
+
+        bytes memory data;
+        data = encodeStorageSetterZeroOutInitializedSlot();
```

Comment:
This approach is pretty tricky as it must understand the underlying storage layout.

One approach that I use is to redeploy both proxy and impl of ASR, and then set the proxy in game impl.  The reason is that we have to upgrade the both game and ASR impls; thus, the extra cost is redeploying the proxy.

Reply:
Yeah, but looks like it's the standard operation as recommended by Mark [here](https://github.com/ethereum-optimism/optimism/issues/11327#issuecomment-2264409100)

Reply:
I see.  Since `storageSetter` is a state-less contract, would `storageSetter` be re-deployed or we can reuse it?

Reply:
Yes, we can reuse if we know the address, but I didn't find the address [here](https://docs.optimism.io/chain/addresses).

Reply:
yes, I mean we may accept a parameter of the address, and if not, deploy one.  In this case, we may re-use mainnet's address if Optimism has one.

Reply:
Good idea, done in 03a3303d60b2959a3566c369d70ed736eb898473 !